app/cluster/state/etcd.go

```diff
@@ -0,0 +1,289 @@
+	AckKey     string `json:"ack_key"`
+}
+
+type workspacesAckValue struct {
+	Status string `json:"status"`
+}
+
+func EnvETCDConfig() *ETCDConfig {
+	endpoints := strings.Split(config.GetEnv("ETCD_HOSTS", "127.0.0.1:2379"), `,`)
+	namespace := config.GetKubeNamespace()
+	serverIndex := config.GetInstanceID()
+
+	envConfigOnce.Do(func() {
+		config.RegisterDurationConfigVariable(time.Duration(15), &etcdGetTimeout, false, time.Second, "etcd.getTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(3), &etcdWatchTimeout, false, time.Second, "etcd.watchTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(30), &keepaliveTime, false, time.Second, "etcd.keepaliveTime")
+		config.RegisterDurationConfigVariable(time.Duration(10), &keepaliveTimeout, false, time.Second, "etcd.keepaliveTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(20), &dialTimeout, false, time.Second, "etcd.dialTimeout")
+	})
+
+	return &ETCDConfig{
+		Endpoints:            endpoints,
+		Namespace:            namespace,
+		ServerIndex:          serverIndex,
+		etcdWatchTimeout:     etcdWatchTimeout,
+		dialTimeout:          dialTimeout,
+		dialKeepAliveTime:    keepaliveTime,
+		dialKeepAliveTimeout: keepaliveTimeout,
+		etcdGetTimeout:       etcdGetTimeout,
+	}
+}
+
+type ETCDManager struct {
+	Config  *ETCDConfig
+	Client  *clientv3.Client
+	once    sync.Once
+	initErr error
+}
+
+func (manager *ETCDManager) init() error {
+	manager.once.Do(func() {
+		cli, err := clientv3.New(clientv3.Config{
+			Endpoints:            manager.Config.Endpoints,
+			DialTimeout:          manager.Config.dialTimeout,
+			DialKeepAliveTime:    manager.Config.dialKeepAliveTime,
+			DialKeepAliveTimeout: manager.Config.dialKeepAliveTimeout,
+			DialOptions: []grpc.DialOption{
+				grpc.WithBlock(), // block until the underlying connection is up
+			},
+		})
+		if err != nil {
+			manager.initErr = err
+			return
+		}
+		manager.Client = cli
+	})
+
+	return manager.initErr
+}
+
+// Ping ensures the connection to etcd is alive
+func (manager *ETCDManager) Ping() error {
+	if err := manager.init(); err != nil {
+		return err
+	}
+
+	_, err := manager.Client.Cluster.MemberList(context.Background())
+	if err != nil {
+		return fmt.Errorf("ping: get cluster member list: %w", err)
+	}
+
+	return nil
+}
+
+func (manager *ETCDManager) unmarshalMode(raw []byte) servermode.ChangeEvent {
+	var req modeRequestValue
+	err := json.Unmarshal(raw, &req)
+	if err != nil {
+		return servermode.ChangeEventError(fmt.Errorf("unmarshal mode request: %w", err))
+	}
+
+	mode := req.Mode
+	if !mode.Valid() {
+		return servermode.ChangeEventError(fmt.Errorf("invalid mode: %s", mode))
+	}
+
+	return servermode.NewChangeEvent(
+		mode,
+		func() error {
+			ackValue, err := json.MarshalToString(modeAckValue{
+				Status: mode,
+			})
+			if err != nil {
+				return fmt.Errorf("marshal ack value: %w", err)
+			}
+			_, err = manager.Client.Put(context.Background(), req.AckKey, ackValue)
```

Comment:
[minor] Perhaps we can convert some of these `context.Background()` to `context.TODO()` :+1:

Reply:
I have passed a proper context here